### PR TITLE
Dumping Pipeline

### PR DESCRIPTION
**Pull request by vijayphoenix**
1. Add new `ToYAML` class and its instances
2. Implement `Data.YAML.Dumper`
3. Some helper functions for testing the output of dumping pipleline in `Main.hs`
4. Move `Data Node` to new module `Data.YAML.Internal`

TODO
1. Integrate Schema handling
2. Integrate more feature to preserve YAML stream
3. Support for failure in `ToYAML`

### CODE REVIEW

File: src/Data/YAML/Dumper.hs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+-- | Dump YAML Nodes using specified UTF encoding to a lazy 'BS.L.ByteString'
+--
+-- @since 0.2.0
+dumpYAML ::  Encoding -> [Node ()] -> BS.L.ByteString
+dumpYAML encoding nodes = writeEvents encoding $ case sequence (dumpEvents nodes) of
+    Left str -> error str
+    Right ev -> ev
+
+-- | Convenience wrapper over 'dumpYAML' expecting exactly one YAML document
```
Comment (vijayphoenix) [inline src/Data/YAML/Dumper.hs]:
```suggestion
-- | Convenience wrapper over 'dumpYAML' expecting exactly one YAML Node
```

File: src-test/Main.hs [hunk 2]
```diff
@@ -474,6 +534,26 @@ cmdTestmlCompiler _ = do
...
 
+printNode :: Node loc -> IO ()
+printNode node = case node of  
+    (Y.Scalar _ a)      -> hPutStrLn stdout $  "Scalar "   ++ show a
+    (Y.Mapping _ a b)   -> do
+                              hPutStrLn stdout $  "Mapping "  ++ show a 
+                              printMap b
+    (Y.Sequence _ a b)  -> do 
+                              hPutStrLn stdout $  "Sequence " ++ show a 
+                              (mapM_ printNode b)
```
Comment (vijayphoenix) [inline src-test/Main.hs]:
```suggestion
                              mapM_ printNode b
```

File: src/Data/YAML/Dumper.hs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        SInt   int    -> Right (YE.Scalar anc (Tag Nothing) Plain (T.pack . show $ int))
+        SStr   text   -> Right (YE.Scalar anc (Tag Nothing) Plain text)
+        SUnknown _ _  -> Left "SUnknown Scalar type"
+
+    goMap :: Int -> Map.Map (Node ()) (Node ()) -> [Node ()] -> Node2EvList -> EvList
+    goMap lvl m rest cont =  goNode lvl (mapToList m) g
+        where 
+          g [] = (Right MappingEnd) : isDocEnd (lvl - 1) rest cont
+          g rest' = goNode lvl rest' g 
+          mapToList = Map.foldrWithKey (\k v a -> k : v : a) []
```
Comment (vijayphoenix) [inline src/Data/YAML/Dumper.hs]: is this fine?

File: src/Data/YAML/Internal.hs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+-- | YAML Document tree/graph
+newtype Doc n = Doc n deriving (Eq,Ord,Show)
+
+-- | YAML mapping
+type Mapping loc = Map (Node loc) (Node loc)
+
+-- | YAML Document node
+data Node loc 
+          = Scalar   !loc !Scalar 
+          | Mapping  !loc !Tag (Map (Node loc) (Node loc))
```
Comment (vijayphoenix) [inline src/Data/YAML/Internal.hs]:
```suggestion
          | Mapping  !loc !Tag (Mapping loc) 
```

File: src/Data/YAML/Dumper.hs [hunk 5]
```diff
@@ -40,49 +40,68 @@ dumpEvents nodes = Right StreamStart: go0 nodes
...
     isDocEnd lvl rest cont = if lvl == 0 then Right (DocumentEnd (rest /= [])): (cont rest) else (cont rest)
 
     ancName :: NodeId -> Maybe Anchor
     ancName (-1) = Nothing
     ancName nid  = Just $ T.pack ("a" ++ show nid)
+
+    dumpBool :: Bool -> T.Text
+    dumpBool b = if b then "true" else "false"
+
+    dumpFloat :: Double -> T.Text
```
Comment (vijayphoenix) [inline src/Data/YAML/Dumper.hs]: `SFloat` is converted to `Text` after making this assumption in [JSON schema](https://github.com/yaml/yaml-spec/issues/29)

File: src/Data/YAML.hs [hunk 6]
```diff
@@ -531,3 +497,80 @@ decode1Strict :: FromYAML v => BS.ByteString -> Either String v
...
+
+
+instance ToYAML Text where
+  toYAML = Scalar () . SStr
+
+instance ToYAML a => ToYAML (Maybe a) where
+  toYAML Nothing = Scalar () SNull
+  toYAML (Just a) = toYAML a
+
+instance (ToYAML a, ToYAML b) => ToYAML (Either a b) where
```
Comment (hvr) [inline src/Data/YAML.hs]: I'd rather not define this instance as it's tricky to round-trip (i.e. we don't have a proper `FromYAML` instance for that either)

File: src/Data/YAML/Dumper.hs [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+import qualified Data.ByteString.Lazy     as BS.L
+import           Data.Map                 as Map
+import qualified Data.Text                as T
+
+
+type EvList = [Either String Event]
+type Node2EvList = [Node ()] -> EvList
+
+-- | Dump YAML Nodes using specified UTF encoding to a lazy 'BS.L.ByteString'
```
Comment (hvr) [inline src/Data/YAML/Dumper.hs]: As I don't think this is obvious it should be explained in the docstring that each Node is emitted as individual "YAML document"

File: src/Data/YAML/Dumper.hs [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+-- @since 0.2.0
+dumpYAML ::  Encoding -> [Node ()] -> BS.L.ByteString
+dumpYAML encoding nodes = writeEvents encoding $ case sequence (dumpEvents nodes) of
+    Left str -> error str
+    Right ev -> ev
+
+-- | Convenience wrapper over 'dumpYAML' expecting exactly one YAML node
+--
+-- @since 0.2.0
+dumpYAML' ::  Encoding -> Node () -> BS.L.ByteString
```
Comment (hvr) [inline src/Data/YAML/Dumper.hs]: This should be the dual of `decodeNode'`

(i.e. this shall take a schema resolver argument; not necessarily this PR; can be done in a future PR)

File: src/Data/YAML.hs [hunk 9]
```diff
@@ -54,8 +54,14 @@
 --
 module Data.YAML
     (
+      -- * Typeclass-based dumping
+      ToYAML(..)
+    , dumpYAML
+    , dumpYAML'
+    , dumpEvents
```
Comment (hvr) [inline src/Data/YAML.hs]: Why is `dumpEvents` exported from this module? The `Data.YAML` module currently does not expose any operation which refers to `Events`

File: src/Data/YAML/Dumper.hs [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+import qualified Data.Text                as T
+
+
+type EvList = [Either String Event]
+type Node2EvList = [Node ()] -> EvList
+
+-- | Dump YAML Nodes using specified UTF encoding to a lazy 'BS.L.ByteString'
+--
+-- @since 0.2.0
+dumpYAML ::  Encoding -> [Node ()] -> BS.L.ByteString
```
Comment (hvr) [inline src/Data/YAML/Dumper.hs]: so this is the dual to `decodeNode`; I'd expect this rather to have a type-sig like

```hs
encodeNode ::  [Doc (Node ())] -> BS.L.ByteString
```

I'd omit the Encoding and default to UTF8, as this is basically supposed to be a convenience function which encodes using the "YAML 1.2 Core schema" and with UTF-8 (which is likely the most common encoding used for YAML files)

File: src/Data/YAML.hs (hunk 9 above)
Comment (vijayphoenix) [reply src/Data/YAML.hs]: ok, will remove that after the debugging process is over

File: src/Data/YAML/Dumper.hs [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+-- | Dump YAML Nodes as a lazy 'BS.L.ByteString'
+--
+-- Each YAML 'Node' is emitted as a individual YAML Document where each Document is terminated by a 'DocumentEnd' indicator.
+--
+-- This is a convenience wrapper over `encodeNode'`
+--
+-- @since 0.2.0
+encodeNode :: [Doc (Node ())] -> BS.L.ByteString
+encodeNode nodes = writeEvents UTF8 $ case sequence (dumpEvents (map getDoc nodes)) of
```
Comment (hvr) [inline src/Data/YAML/Dumper.hs]: I'm not happy with this `case sequence ... of` but we'll deal with this in a different PR; I'm going to merge this PR now

Comment (vijayphoenix) [reply src/Data/YAML/Dumper.hs]: Resolved, see PR #21

**Pull request merged**